--- ../src-base/minecraft/net/minecraft/server/management/PlayerChunkMapEntry.java
+++ ../src-work/minecraft/net/minecraft/server/management/PlayerChunkMapEntry.java
@@ -19,25 +19,34 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import carpet.CarpetSettings;
+import carpet.carpetclient.CarpetClientChunkLogger;
+import carpet.helpers.IPlayerSensitiveTileEntity;
+
 public class PlayerChunkMapEntry
 {
     private static final Logger field_187281_a = LogManager.getLogger();
     private final PlayerChunkMap field_187282_b;
-    private final List<EntityPlayerMP> field_187283_c = Lists.<EntityPlayerMP>newArrayList();
+    public final List<EntityPlayerMP> field_187283_c = Lists.<EntityPlayerMP>newArrayList(); // CM: changed to public
+    private final List<EntityPlayerMP> renderingplayers = Lists.<EntityPlayerMP>newArrayList();
     private final ChunkPos field_187284_d;
     private final short[] field_187285_e = new short[64];
     @Nullable
-    private Chunk field_187286_f;
+    public Chunk field_187286_f; // CM: changed to public
     private int field_187287_g;
     private int field_187288_h;
     private long field_187289_i;
-    private boolean field_187290_j;
+    public boolean field_187290_j; // CM: changed to public
 
-    public PlayerChunkMapEntry(PlayerChunkMap p_i1518_1_, int p_i1518_2_, int p_i1518_3_)
+    public PlayerChunkMapEntry(PlayerChunkMap mapIn, int chunkX, int chunkZ, EntityPlayerMP player)
     {
-        this.field_187282_b = p_i1518_1_;
-        this.field_187284_d = new ChunkPos(p_i1518_2_, p_i1518_3_);
-        this.field_187286_f = p_i1518_1_.func_72688_a().func_72863_F().func_186028_c(p_i1518_2_, p_i1518_3_);
+        this.field_187282_b = mapIn;
+        this.field_187284_d = new ChunkPos(chunkX, chunkZ);
+        if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+            carpet.carpetclient.CarpetClientChunkLogger.setReason("Player loading chunk");
+        // Added a way to remove spectators loading chunks. CARPET-XCOM
+        if(!CarpetSettings.spectatorsDontLoadChunks || !player.func_175149_v()) this.field_187286_f = mapIn.func_72688_a().func_72863_F().func_186028_c(chunkX, chunkZ);
+        carpet.carpetclient.CarpetClientChunkLogger.resetReason();
     }
 
     public ChunkPos func_187264_a()
@@ -53,36 +62,75 @@
         }
         else
         {
+            // ChunkLogger - 0x-CARPET
+            if(CarpetClientChunkLogger.logger.enabled) {
+                if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+                    carpet.carpetclient.CarpetClientChunkLogger.setReason("Player added to chunk");
+            	CarpetClientChunkLogger.logger.log(this.field_187282_b.func_72688_a(),field_187284_d.field_77276_a,field_187284_d.field_77275_b,CarpetClientChunkLogger.Event.PLAYER_ENTERS);
+                carpet.carpetclient.CarpetClientChunkLogger.resetReason();
+            }
+
             if (this.field_187283_c.isEmpty())
             {
                 this.field_187289_i = this.field_187282_b.func_72688_a().func_82737_E();
             }
 
             this.field_187283_c.add(p_187276_1_);
+        }
 
+        // Added to enable randomticking chunk updates Carpet-XCOM
+        carpet.carpetclient.CarpetClientRandomtickingIndexing.enableUpdate(p_187276_1_);
+    }
+
+    public void addRenderingPlayer(EntityPlayerMP player)
+    {
+        if (this.renderingplayers.contains(player))
+        {
+            field_187281_a.debug("Failed to add player. {} already is in chunk {}, {}", player, Integer.valueOf(this.field_187284_d.field_77276_a), Integer.valueOf(this.field_187284_d.field_77275_b));
+        }
+        else
+        {
+            this.renderingplayers.add(player);
+
             if (this.field_187290_j)
             {
-                this.func_187278_c(p_187276_1_);
+                this.func_187278_c(player);
             }
         }
     }
 
     public void func_187277_b(EntityPlayerMP p_187277_1_)
     {
-        if (this.field_187283_c.contains(p_187277_1_))
+        if (this.renderingplayers.contains(p_187277_1_))
         {
             if (this.field_187290_j)
             {
                 p_187277_1_.field_71135_a.func_147359_a(new SPacketUnloadChunk(this.field_187284_d.field_77276_a, this.field_187284_d.field_77275_b));
             }
 
+            this.renderingplayers.remove(p_187277_1_);
+        }
+
+        if (this.field_187283_c.contains(p_187277_1_))
+        {
             this.field_187283_c.remove(p_187277_1_);
 
             if (this.field_187283_c.isEmpty())
             {
+                // ChunkLogger - 0x-CARPET
+                if(CarpetClientChunkLogger.logger.enabled) {
+                    if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+                        carpet.carpetclient.CarpetClientChunkLogger.setReason("Player removed from chunk");
+                	CarpetClientChunkLogger.logger.log(this.field_187282_b.func_72688_a(),field_187284_d.field_77276_a,field_187284_d.field_77275_b,CarpetClientChunkLogger.Event.PLAYER_LEAVES);
+                    carpet.carpetclient.CarpetClientChunkLogger.resetReason();
+                }
+
                 this.field_187282_b.func_187305_b(this);
             }
         }
+
+        // Added to enable randomticking chunk updates Carpet-XCOM
+        carpet.carpetclient.CarpetClientRandomtickingIndexing.enableUpdate(p_187277_1_);
     }
 
     public boolean func_187268_a(boolean p_187268_1_)
@@ -95,7 +143,10 @@
         {
             if (p_187268_1_)
             {
+                if(carpet.carpetclient.CarpetClientChunkLogger.logger.enabled)
+                    carpet.carpetclient.CarpetClientChunkLogger.setReason("Player loading new chunks and generating");
                 this.field_187286_f = this.field_187282_b.func_72688_a().func_72863_F().func_186025_d(this.field_187284_d.field_77276_a, this.field_187284_d.field_77275_b);
+                carpet.carpetclient.CarpetClientChunkLogger.resetReason();
             }
             else
             {
@@ -127,7 +178,7 @@
             this.field_187290_j = true;
             Packet<?> packet = new SPacketChunkData(this.field_187286_f, 65535);
 
-            for (EntityPlayerMP entityplayermp : this.field_187283_c)
+            for (EntityPlayerMP entityplayermp : this.renderingplayers)
             {
                 entityplayermp.field_71135_a.func_147359_a(packet);
                 this.field_187282_b.func_72688_a().func_73039_n().func_85172_a(entityplayermp, this.field_187286_f);
@@ -190,9 +241,9 @@
     {
         if (this.field_187290_j)
         {
-            for (int i = 0; i < this.field_187283_c.size(); ++i)
+            for (int i = 0; i < this.renderingplayers.size(); ++i)
             {
-                (this.field_187283_c.get(i)).field_71135_a.func_147359_a(p_187267_1_);
+                (this.renderingplayers.get(i)).field_71135_a.func_147359_a(p_187267_1_);
             }
         }
     }
@@ -248,12 +299,28 @@
     {
         if (p_187273_1_ != null)
         {
+            if (p_187273_1_ instanceof IPlayerSensitiveTileEntity)
+            {
+                if (field_187290_j)
+                {
+                    for (int i = 0; i < renderingplayers.size(); i++)
+                    {
+                        EntityPlayerMP player = renderingplayers.get(i);
+                        SPacketUpdateTileEntity packet = ((IPlayerSensitiveTileEntity) p_187273_1_).getUpdatePacketPlayerSensitive(player);
+                        if (packet != null)
+                            player.field_71135_a.func_147359_a(packet);
+                    }
+                }
+            }
+            else
+            {
             SPacketUpdateTileEntity spacketupdatetileentity = p_187273_1_.func_189518_D_();
 
             if (spacketupdatetileentity != null)
             {
                 this.func_187267_a(spacketupdatetileentity);
             }
+            }
         }
     }
 
@@ -262,6 +329,11 @@
         return this.field_187283_c.contains(p_187275_1_);
     }
 
+    public boolean containsRenderingPlayer(EntityPlayerMP player)
+    {
+        return this.renderingplayers.contains(player);
+    }
+
     public boolean func_187269_a(Predicate<EntityPlayerMP> p_187269_1_)
     {
         return Iterables.tryFind(this.field_187283_c, p_187269_1_).isPresent();
